Update lint rule name & implement rendering mazes as SVGs & more algorithms

langp11 opened the pull request:
Running cargo test generated this warning:
```
 --> D:\Data\Dev\rust\projects\maze_generator\src\lib.rs:3:5
  |
3 |     missing_crate_level_docs,
  |     ^^^^^^^^^^^^^^^^^^^^^^^^ help: use the new name: `rustdoc::missing_crate_level_docs`
  |
  = note: `#[warn(renamed_and_removed_lints)]` on by default
```

ftsell reviewed (changes requested):
Thanks for the contribution :)
The first commit about the old lint name I will definitely accept as it is.

The second feature I am also positively inclined to accept as it doesn't look too complex and offers a very nice visualization.
However I have some remarks which I would like to get addressed first:
- You have implemented the rendering options as a HashMap which does work but I generally prefer things like these to have their own concrete type. That way a user can clearly infer what options are available and what they do since they can be documented on the type. I would then also like this options type to implement `Default` instead of having the parameter passed as an `Option`.
- ~As a user I would expect a method named `to_svg` to return some sort of SVG (maybe as a String) instead of simply writing it to stdout. This also increases the flexibility in using the function. Alternatively you could let the method accept a `impl std::io::Write` or some other buffer parameter into which the resulting SVG gets written.~ Nevermind, I overlooked that you already do that

langp11 commented:
Yes, the struct approach is clearly better than the HashMap and I have made changes accordingly.
I'm not sure if I've made the argument passing work as you describe - the example shows the new mechanism with both default and (some) specified options.

I've also added some new algorithms - these might belong as a separate PR, but my git skills are not what they should be...

ftsell commented:
> these might belong as a separate PR, but my git skills are not what they should be...

I kinda agree although it is fine to keep it in one PR for now since there isn't much other activity on the repo.
In case you wanna learn, I recommend reading [Learn to change history with git rebase](https://git-rebase.io/) and then perform the following steps:
1. Create one or two separate branches from your current branch. These should each contain one logical patch (i.e. one feature)
2. Remove the commits which are not related to the patch from each branch using `git rebase`
3. Create a separate pull request for each of these branches.

You might have some difficulty if the patches depend on each other in which case you could either submit the second one only after the first one has been merged or keep them in the same patch.

---

Regarding the review, I'll try to get to it on the weekend :)

lilioid commented on Cargo.toml:
```diff
@@ -20,3 +20,7 @@ petgraph = "0.5.1"
+clap = { version = "3.1.5", features = ["derive"] }
+
+[[example]]
+name="mazes"
```
You don't actually need to manually list this example since cargo is able auto-discover examples. 

See [reference about target auto-discovery](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#target-auto-discovery)

lilioid commented on src/prelude/svgoptions.rs:
```diff
@@ -0,0 +1,42 @@
+use crate::prelude::*;
```
This is an unused import. Please remove it.

lilioid commented on src/prelude/svgoptions.rs:
```diff
@@ -0,0 +1,42 @@
+use crate::prelude::*;
+use std::fmt::{Debug, Display, Formatter};
```
`Display` and `Formatter` are unused imports as well. Please remove them.

lilioid commented on src/prelude/svgoptions.rs:
```diff
@@ -0,0 +1,42 @@
+use crate::prelude::*;
+use std::fmt::{Debug, Display, Formatter};
+
+/// Otions for enerating SVG output
```
Please fix the typos

lilioid commented on src/prelude/svgoptions.rs:
```diff
@@ -0,0 +1,42 @@
+
+/// Otions for enerating SVG output
+#[derive(Debug)]
+pub struct SVGoptions {
```
Proper CamelCase would be either `SvgOptions` or `SVGOptions` although I prefer the first. Please use one of the two.

lilioid commented on src/prelude/svgoptions.rs:
```diff
@@ -0,0 +1,42 @@
+    /// Padding, default: 10
+    pub padding: i32,
+    /// Height in pixels - use zero to derive a height based on the padding
+    pub height: i32,
```
I personally don't like treating zeros as a special *automatic* value.
I would prefer using an `Option<i32>` to explicitly treat it as an unspecified value.

lilioid commented on src/prelude/svgoptions.rs:
```diff
@@ -0,0 +1,42 @@
+impl SVGoptions {
+    /// Create a new instance from the specified coordinate components
+    pub fn new() -> Self {
+        Default::default()
```
The docstring here is wrong since we are not creating anything from coordinate components but just using the default.

Also, I would accept parameters in the `new` function since a user can always call `default()` themselves if they don't want to specify anything special.
In case the user only wants to overwrite partial settings, you could maybe implement a [builder pattern](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html) for the `SvgOptions` too.

lilioid commented on src/growing_tree.rs:
```diff
@@ -0,0 +1,165 @@
+    /// Optionally a 32 bit seed can be provided to seed the internal random generator.
+    /// Giving a seed results in identical mazes being generated which omitting it sources the
+    /// random generator from entropy.
+    ///
+    ///
```
Please remove the two extraneous and unused lines

lilioid commented on src/growing_tree.rs:
```diff
@@ -0,0 +1,165 @@
+pub struct GrowingTreeGenerator {
+    rng: ChaChaRng,
+    selectionmethod: i32,
+    cellstack: Vec<Coordinates>,
```
Please use snake_case for `selection_method` and `call_stack`

lilioid commented on src/growing_tree.rs:
```diff
@@ -0,0 +1,165 @@
+    rng: ChaChaRng,
+    selectionmethod: i32,
+    cellstack: Vec<Coordinates>,
+    //frontier: Vec<Coordinates>,
```
If this is just left over from development and not needed anymore, please remove it

lilioid commented on src/growing_tree.rs:
```diff
@@ -0,0 +1,165 @@
+#[derive(Debug, Clone)]
+pub struct GrowingTreeGenerator {
+    rng: ChaChaRng,
+    selectionmethod: i32,
```
You restrict the selection method to three possible value types (as described in `set_selectionmethod` and later seen in code).
Please use an enum for specifying the selection method since that is one of the things they were designed for. You can then also directly describe the selection method on the enum member in detail instead of on the setter method.

The enum could look like the following:
```rust
pub enum GrowingTreeSelectionMethod {
    /// <some description>
    Random,
    /// <some description>
    MostRecent,
    /// <some description>
    First,
}
```

lilioid commented on src/growing_tree.rs:
```diff
@@ -0,0 +1,165 @@
+    /// any other value = First. This is the constructor default.
+    ///
+    /// This is a setter method so that all the generator constructors retain the same signature
+    pub fn set_selectionmethod(&mut self, selectionmethod: i32) {
```
In this specific case, the field `selection_method` on the struct could just simply be public.
That way a user can also read the current selection method and we don't need a setter.

lilioid commented on src/growing_tree.rs:
```diff
@@ -0,0 +1,165 @@
+        goal_coordinates
+    }
+
+    // Find the neighbours of this cell that have NOT been visited
```
Please use a proper docstring with three `/` characters

lilioid commented on src/growing_tree.rs:
```diff
@@ -0,0 +1,165 @@
+
+        while ! self.cellstack.is_empty() {
+            self.find_unvisited_neighbours(maze, current_coordinates);
+            //eprintln!("Stacks: cellstack {}, neighbours {}, current {:?}", self.cellstack.len(), self.neighbours.len(), current_coordinates);
```
If this debug print is not needed anymore, please remove it

lilioid commented on src/prims_algorithm.rs:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    // Mark a cell as visited and it's unvisited neighbours as frontier cells
+    fn markcell(&mut self, maze: &mut Maze, current_coordinates: Coordinates) {
```
Please use snake_case so `mark_cell`

lilioid commented on src/prims_algorithm.rs:
```diff
@@ -0,0 +1,164 @@
+        }
+    }
+
+    // Find the neighbours of this cell that have been visited
```
Please use three `/` characters for a proper docstring

lilioid commented on src/prims_algorithm.rs:
```diff
@@ -0,0 +1,164 @@
+
+    // do breadth-first search for the field which has the most distance
+    // Cloned from ellers_algorithm, but passing in maze rather than using the self.graph element (which we don't have here)
+    fn find_suitable_goal(&self, maze: &mut Maze, start: Coordinates) -> Coordinates {
```
Please use three `/` characters for a proper docstring

lilioid commented on src/prelude/maze.rs:
```diff
@@ -116,6 +118,145 @@ impl std::fmt::Debug for Maze {
 }
 
+impl Maze {
+    /// Generate an SVG version of the maze, returned as a String which you can then write to a file...or whatever
```
The "or whatever" is a bit too informal in my opinion

lilioid commented on examples/mazes.rs:
```diff
@@ -0,0 +1,171 @@
+        ..Default::default()
+    };
+
+    if svgoutput {
```
You probably want to remove this line because it is a duplicate

lilioid commented on examples/mazes.rs:
```diff
@@ -0,0 +1,171 @@
+    // Define the CLI arguments
+    let matches = Command::new("Maze Test Program")
+        .version("0.1.0")
+        .about("Test harness for maze_generator library crate")
```
Please change this to something that explains that this is an example binary. It can be a short explanation, that's fine. It's just not a test harness.

ftsell reviewed (changes requested):
Alright, the code is mostly alright and I definately appreciate the new features. Especially the SVG output I find really useful but the new generators are fun too :)

I do have some suggestions thoug most of them are just coding-style hints. Nonetheless please have a look at the review comments and address them.

langp11 commented:
Hi @ftsell, I think I've addressed all your suggestions (which all seem sensible and 'rusty'), with the exception of the "I would accept parameters in the new function..." bit. I can see why it would be a Good Thing, I'm just not sure how to implement it due to a lack of rust-fu on my part.

ftsell reviewed (approved):
Alright, good job.
I'll merge this now, change and add some other minor things, and then do a release later in the day.

Thanks again for the contribution :)
